pkg/util/ssh.go

```diff
@@ -239,95 +243,186 @@ func ParsePublicKeyFromFile(keyFile string) (*rsa.PublicKey, error) {
+}
+
+type sshTunnelEntry struct {
 	Address string
-	Tunnel  *SSHTunnel
+	Tunnel  tunnel
+}
+
+type sshTunnelCreator interface {
+	NewSSHTunnel(user, keyFile, healthCheckPath string) (tunnel, error)
+}
+
+type realTunnelCreator struct{}
+
+func (*realTunnelCreator) NewSSHTunnel(user, keyFile, healthCheckPath string) (tunnel, error) {
+	return NewSSHTunnel(user, keyFile, healthCheckPath)
 }
 
-// Not thread safe!
 type SSHTunnelList struct {
-	entries []SSHTunnelEntry
+	entries       []sshTunnelEntry
+	adding        map[string]bool
+	tunnelCreator sshTunnelCreator
+	tunnelsLock   sync.Mutex
+
+	user            string
+	keyfile         string
+	healthCheckPath string
 }
 
-func MakeSSHTunnels(user, keyfile string, addresses []string) *SSHTunnelList {
-	tunnels := []SSHTunnelEntry{}
-	for ix := range addresses {
-		addr := addresses[ix]
-		tunnel, err := NewSSHTunnel(user, keyfile, addr)
-		if err != nil {
-			glog.Errorf("Failed to create tunnel for %q: %v", addr, err)
-			continue
-		}
-		tunnels = append(tunnels, SSHTunnelEntry{addr, tunnel})
+func NewSSHTunnelList(user, keyfile, healthCheckPath string, stopChan chan struct{}) *SSHTunnelList {
+	l := &SSHTunnelList{
+		adding:          make(map[string]bool),
+		tunnelCreator:   &realTunnelCreator{},
+		user:            user,
+		keyfile:         keyfile,
+		healthCheckPath: healthCheckPath,
 	}
-	return &SSHTunnelList{tunnels}
+	healthCheckPoll := 1 * time.Minute
+	go Until(func() {
+		l.tunnelsLock.Lock()
+		defer l.tunnelsLock.Unlock()
+		// Healthcheck each tunnel every minute
+		numTunnels := len(l.entries)
+		for i, entry := range l.entries {
+			sleepTime := healthCheckPoll * time.Duration(i) / time.Duration(numTunnels)
+			entry := entry
+			go func() {
+				time.Sleep(sleepTime)
+				if err := l.healthCheck(entry); err != nil {
+					glog.Errorf("Healthcheck failed for tunnel to %q: %v", entry.Address, err)
+					glog.Infof("Attempting once to re-establish tunnel to %q", entry.Address)
+					l.removeAndReAdd(entry)
+				}
+			}()
+		}
+	}, healthCheckPoll, stopChan)
+	return l
 }
 
-// Open attempts to open all tunnels in the list, and removes any tunnels that
-// failed to open.
-func (l *SSHTunnelList) Open() error {
-	var openTunnels []SSHTunnelEntry
-	for ix := range l.entries {
-		if err := l.entries[ix].Tunnel.Open(); err != nil {
-			glog.Errorf("Failed to open tunnel %v: %v", l.entries[ix], err)
-		} else {
-			openTunnels = append(openTunnels, l.entries[ix])
-		}
-	}
-	l.entries = openTunnels
-	if len(l.entries) == 0 {
-		return errors.New("Failed to open any tunnels.")
-	}
-	return nil
+func (l *SSHTunnelList) healthCheck(e sshTunnelEntry) error {
+	// GET the healthcheck path using the provided tunnel's dial function.
+	transport := utilnet.SetTransportDefaults(&http.Transport{
+		Dial:            e.Tunnel.Dial,
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	})
+	client := &http.Client{Transport: transport}
+	_, err := client.Get(l.healthCheckPath)
```

Comment:
Checking only the error just checks that we could get a response (so not really a true kubelet health check). For tunnel health, I think this is enough.